File: examples/gemini/langgraph_agent/graph.py
"""LangGraph ReAct agent powered by Gemini (real LLM).

Requires GOOGLE_API_KEY (see repo-root .env.llm / .env.llm.example).
Deterministic tools; non-deterministic model text — use for live N×N
validation, not for golden event fixtures.
"""

from __future__ import annotations

import sys
from pathlib import Path
from typing import Annotated, TypedDict

from langchain_core.messages import BaseMessage
from langchain_core.tools import tool
from langchain_google_genai import ChatGoogleGenerativeAI
from langgraph.graph import END, START, StateGraph
from langgraph.graph.message import add_messages
from langgraph.prebuilt import ToolNode, tools_condition

sys.path.insert(0, str(Path(__file__).resolve().parents[1]))
from _env import gemini_model, gemini_temperature, require_google_api_key  # noqa: E402


@tool
def search(query: str) -> str:
    """Search for factual information about a topic."""
    return f"Result for '{query}': Runkite is an Agent Protocol control plane with pluggable runners."


class State(TypedDict):
    messages: Annotated[list[BaseMessage], add_messages]


_llm = ChatGoogleGenerativeAI(
    model=gemini_model(),
    google_api_key=require_google_api_key(),
    temperature=gemini_temperature(),
).bind_tools([search])

_tools = ToolNode([search])


def agent_node(state: State):
    return {"messages": [_llm.invoke(state["messages"])]}


builder = StateGraph(State)
builder.add_node("agent", agent_node)
builder.add_node("tools", _tools)
builder.add_edge(START, "agent")
builder.add_conditional_edges("agent", tools_condition)
builder.add_edge("tools", "agent")
builder.add_edge("agent", END)

graph = builder.compile()

File: examples/gemini/langgraphjs_agent/graph.ts
/**
 * LangGraph.js agent powered by Gemini (real LLM).
 * Requires GOOGLE_API_KEY in the environment (load from repo-root .env.llm).
 *
 * Returns the model AIMessage (with usage_metadata) into state — not a
 * stripped {role,content} dict — so the runner's accumulateUsage can meter
 * FinOps the same way as the Python LangGraph example.
 */
import { ChatGoogleGenerativeAI } from "@langchain/google-genai";
import { END, MessagesAnnotation, START, StateGraph } from "@langchain/langgraph";

// MessagesAnnotation (LangGraph.js's equivalent of Python's
// Annotated[list, add_messages]) instead of a bare `.concat()` reducer:
// besides the usual message-list conveniences (merge-by-id, RemoveMessage
// support), it stamps a stable id onto every message that lacks one. The
// runner's FinOps usage-dedup (executeRun.ts's pre-run getState snapshot)
// keys off that id to tell "already billed in an earlier turn" apart from
// "new this run" -- a plain concat reducer leaves messages id-less, so
// every turn on a thread would look "new" and get summed all over again.
const State = MessagesAnnotation;

const modelName = process.env.GEMINI_MODEL || "gemini-2.0-flash";
const temperature = Number(process.env.GEMINI_TEMPERATURE || "0");

const llm = new ChatGoogleGenerativeAI({
  model: modelName,
  apiKey: process.env.GOOGLE_API_KEY,
  temperature,
});

async function agentNode(state: typeof State.State) {
  const last = state.messages[state.messages.length - 1];
  const lastContent = typeof last.content === "string" ? last.content : JSON.stringify(last.content);
  const resp = await llm.invoke([
    { role: "system", content: "You are a terse, helpful assistant. Reply in one short sentence." },
    { role: "user", content: lastContent },
  ]);
  // ChatGoogleGenerativeAI returns usage_metadata tokens but omits model
  // from response_metadata — stamp it so pricebook soft-match can price USD.
  resp.response_metadata = {
    ...(resp.response_metadata ?? {}),
    model_name: modelName,
  };
  // Keep the AIMessage intact (usage_metadata / response_metadata) for FinOps.
  return { messages: [resp] };
}

const builder = new StateGraph(State)
  .addNode("agent", agentNode)
  .addEdge(START, "agent")
  .addEdge("agent", END);

export const graph = builder.compile();
